gqcp/include/QCMethod/HF/UHF/UHFSCFSolver.hpp

```diff
@@ -69,16 +69,10 @@ class UHFSCFSolver {
             .add(UHFElectronicEnergyCalculation<Scalar>());
 
         // Create a compound convergence criterion on the norm of subsequent alpha- and beta-density matrices
-        using SingleConvergenceType = ConsecutiveIteratesNormConvergence<OneDM<Scalar>, UHFSCFEnvironment<Scalar>>;
-
-        const auto density_matrix_alpha_extractor = [](const UHFSCFEnvironment<Scalar>& environment) { return environment.density_matrices_alpha; };
-        const SingleConvergenceType convergence_criterion_alpha {threshold, density_matrix_alpha_extractor, "the UHF alpha-density matrix in AO basis"};
-
-        const auto density_matrix_beta_extractor = [](const UHFSCFEnvironment<Scalar>& environment) { return environment.density_matrices_beta; };
-        const SingleConvergenceType convergence_criterion_beta {threshold, density_matrix_beta_extractor, "the UHF beta-density matrix in AO basis"};
-
-        const CompoundConvergenceCriterion<UHFSCFEnvironment<Scalar>> convergence_criterion {convergence_criterion_alpha, convergence_criterion_beta};
+        using SingleConvergenceType = ConsecutiveIteratesNormConvergence<SpinResolvedOneDM<Scalar>, UHFSCFEnvironment<Scalar>>;
 
+        const auto density_matrix_extractor = [](const UHFSCFEnvironment<Scalar>& environment) { return environment.density_matrices; };
+        const SingleConvergenceType convergence_criterion {threshold, density_matrix_extractor, "the UHF density matrices in AO basis"};
```

Comment:
I see that you're now using the `.norm`-method that you've implemented on `SpinResolvedOneDM, which is the reason you're reporting convergence problems.

Reply:
Yeah, I already figured something was going wrong there. I was looking into it already. I tried to implement it more similar to how it first was, with a separate lambda function to extract the alpha and beta density matrices, but that didn't work either.

Reply:
I discovered some additional info while trying to resolve this problem. 

-  When using the code above, with the (wrong) `.norm()` function, the method does converge, but it converges wrongly. (I think it stops too early, leading to an energy estimate that's not low enough). 

- As mentioned, I also tried another method, more similarly implemented to the original code. However this causes the algorithm to stop converging altogether. I've added the code below in order to show you my thought process.

```C++
const auto density_matrix_alpha_extractor = [](const UHFSCFEnvironment<Scalar>& environment) {
            std::deque<GQCP::OneDM<Scalar>> alpha_density_matrices;

            for (size_t i; i < environment.density_matrices.size(); i++) {
                alpha_density_matrices.push_back(environment.density_matrices[i].alpha());
            }
            return alpha_density_matrices;
        };

        const auto density_matrix_beta_extractor = [](const UHFSCFEnvironment<Scalar>& environment) {
            std::deque<GQCP::OneDM<Scalar>> beta_density_matrices;

            for (size_t i; i < environment.density_matrices.size(); i++) {
                beta_density_matrices.push_back(environment.density_matrices[i].beta());
            }
            return beta_density_matrices;
        };

        const SingleConvergenceType convergence_criterion_alpha {threshold, density_matrix_alpha_extractor, "the UHF alpha-density matrix in AO basis"};
        const SingleConvergenceType convergence_criterion_beta {threshold, density_matrix_beta_extractor, "the UHF beta-density matrix in AO basis"};
        const CompoundConvergenceCriterion<UHFSCFEnvironment<Scalar>> convergence_criterion {convergence_criterion_alpha, convergence_criterion_beta};
```

Reply:
EDIT 2: Disccovered the mistake in my loops and this method should work fine now.